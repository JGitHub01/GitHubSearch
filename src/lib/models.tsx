/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

export interface ISearchQuery {
  keyword?: string
  language?: string
  sort?: string
  order?: string
  perPage?: number
  page?: number
}
export interface IRepositoryOwner {
  login: string,
  id: number,
  avatar_url: string,
  url: string,
  html_url: string
}
interface IMatch {
  text: string
  indices: number[]
}
interface ITextMatch {
  property: string
  fragment: string
  matches: IMatch[]
}
export interface IRepository {
  id: number,
  name: string,
  full_name: string,
  owner: IRepositoryOwner,
  description: string,
  url: string,
  html_url: string,
  stargazers_count: number,
  language: string
  text_matches?: ITextMatch[]
}
export interface ISearchResult {
  total_count: number
  incomplete_results?: boolean
  items?: IRepository[]
  success?: boolean
  msg?: string
}
