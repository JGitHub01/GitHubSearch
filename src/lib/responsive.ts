const screenMap = {
  'sm': '639px',
  'md': '768px',
  'lg': '992px'
}

export type ScreenSize = 'sm' | 'md' | 'lg';

export function isScreenWidthAtLeast(size: ScreenSize) {
  const minWidth = screenMap[size];
  return window.matchMedia(`(min-width: ${minWidth})`).matches;
}
