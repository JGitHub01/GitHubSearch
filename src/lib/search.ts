import { ISearchQuery, ISearchResult } from 'lib/models';

export const gitHubApiUrl = 'https://api.github.com/search/repositories';
export default async function search(query: ISearchQuery): Promise<ISearchResult> {
  const {
    keyword = '',
    language = 'javascript',
    sort = '',
    order = 'desc',
    perPage = 20,
    page = 1
  } = query;
  const queryStr = new URLSearchParams({
    q: `${keyword}+language:${language}`,
    sort: sort,
    order: order,
    per_page: perPage.toString(),
    page: page.toString()
  }).toString();
  const searchUrl = `${gitHubApiUrl}?${queryStr}`;
  try {
    const response = await fetch(searchUrl, {
      method: 'GET',
    });
    if (response.status === 200) {
      return response.json();
    } else {
      return {
        total_count: 0,
        success: false,
        msg: response.status.toString()
      };
    }
  } catch (e) {
    return {
      total_count: 0,
      success: false,
      msg: e.toString()
    };
  }
}
