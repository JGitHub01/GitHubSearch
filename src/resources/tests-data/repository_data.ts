export const search_result = {
  "keyword": "tetri",
  "isSearching": true,
  "total_count": 40,
  "incomplete_results": false,
  "items": [
    {
      "id": 343937425,
      "node_id": "MDEwOlJlcG9zaXRvcnkzNDM5Mzc0MjU=",
      "name": "Tetrio-Rank-Predictor",
      "full_name": "chunkanglu/Tetrio-Rank-Predictor",
      "private": false,
      "owner": {
        "login": "chunkanglu",
        "id": 75538064,
        "node_id": "MDQ6VXNlcjc1NTM4MDY0",
        "avatar_url": "https://avatars.githubusercontent.com/u/75538064?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/chunkanglu",
        "html_url": "https://github.com/chunkanglu",
        "followers_url": "https://api.github.com/users/chunkanglu/followers",
        "following_url": "https://api.github.com/users/chunkanglu/following{/other_user}",
        "gists_url": "https://api.github.com/users/chunkanglu/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/chunkanglu/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/chunkanglu/subscriptions",
        "organizations_url": "https://api.github.com/users/chunkanglu/orgs",
        "repos_url": "https://api.github.com/users/chunkanglu/repos",
        "events_url": "https://api.github.com/users/chunkanglu/events{/privacy}",
        "received_events_url": "https://api.github.com/users/chunkanglu/received_events",
        "type": "User",
        "site_admin": false
      },
      "html_url": "https://github.com/chunkanglu/Tetrio-Rank-Predictor",
      "description": "Provides an approximate rank for Tetr.io's Tetra League based on APM, PPS, and VS-Score.",
      "fork": false,
      "url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor",
      "forks_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/forks",
      "keys_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/keys{/key_id}",
      "collaborators_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/collaborators{/collaborator}",
      "teams_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/teams",
      "hooks_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/hooks",
      "issue_events_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/issues/events{/number}",
      "events_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/events",
      "assignees_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/assignees{/user}",
      "branches_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/branches{/branch}",
      "tags_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/tags",
      "blobs_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/git/blobs{/sha}",
      "git_tags_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/git/tags{/sha}",
      "git_refs_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/git/refs{/sha}",
      "trees_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/git/trees{/sha}",
      "statuses_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/statuses/{sha}",
      "languages_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/languages",
      "stargazers_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/stargazers",
      "contributors_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/contributors",
      "subscribers_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/subscribers",
      "subscription_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/subscription",
      "commits_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/commits{/sha}",
      "git_commits_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/git/commits{/sha}",
      "comments_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/comments{/number}",
      "issue_comment_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/issues/comments{/number}",
      "contents_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/contents/{+path}",
      "compare_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/compare/{base}...{head}",
      "merges_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/merges",
      "archive_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/{archive_format}{/ref}",
      "downloads_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/downloads",
      "issues_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/issues{/number}",
      "pulls_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/pulls{/number}",
      "milestones_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/milestones{/number}",
      "notifications_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/notifications{?since,all,participating}",
      "labels_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/labels{/name}",
      "releases_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/releases{/id}",
      "deployments_url": "https://api.github.com/repos/chunkanglu/Tetrio-Rank-Predictor/deployments",
      "created_at": "2021-03-02T23:04:08Z",
      "updated_at": "2021-03-08T20:29:00Z",
      "pushed_at": "2021-03-08T20:28:58Z",
      "git_url": "git://github.com/chunkanglu/Tetrio-Rank-Predictor.git",
      "ssh_url": "git@github.com:chunkanglu/Tetrio-Rank-Predictor.git",
      "clone_url": "https://github.com/chunkanglu/Tetrio-Rank-Predictor.git",
      "svn_url": "https://github.com/chunkanglu/Tetrio-Rank-Predictor",
      "homepage": null,
      "size": 9015,
      "stargazers_count": 0,
      "watchers_count": 0,
      "language": "Python",
      "has_issues": true,
      "has_projects": true,
      "has_downloads": true,
      "has_wiki": true,
      "has_pages": false,
      "forks_count": 0,
      "mirror_url": null,
      "archived": false,
      "disabled": false,
      "open_issues_count": 0,
      "license": null,
      "forks": 0,
      "open_issues": 0,
      "watchers": 0,
      "default_branch": "main",
      "score": 1
    },
    {
      "id": 324767424,
      "node_id": "MDEwOlJlcG9zaXRvcnkzMjQ3Njc0MjQ=",
      "name": "Cooperative-Protein-Dynamics",
      "full_name": "research555/Cooperative-Protein-Dynamics",
      "private": false,
      "owner": {
        "login": "research555",
        "id": 62707744,
        "node_id": "MDQ6VXNlcjYyNzA3NzQ0",
        "avatar_url": "https://avatars.githubusercontent.com/u/62707744?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/research555",
        "html_url": "https://github.com/research555",
        "followers_url": "https://api.github.com/users/research555/followers",
        "following_url": "https://api.github.com/users/research555/following{/other_user}",
        "gists_url": "https://api.github.com/users/research555/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/research555/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/research555/subscriptions",
        "organizations_url": "https://api.github.com/users/research555/orgs",
        "repos_url": "https://api.github.com/users/research555/repos",
        "events_url": "https://api.github.com/users/research555/events{/privacy}",
        "received_events_url": "https://api.github.com/users/research555/received_events",
        "type": "User",
        "site_admin": false
      },
      "html_url": "https://github.com/research555/Cooperative-Protein-Dynamics",
      "description": "This is a simple program that maps the dynamics between LacI and TetR in a one of the first toggle switch systems created by Gardner et al (2000)",
      "fork": false,
      "url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics",
      "forks_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/forks",
      "keys_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/keys{/key_id}",
      "collaborators_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/collaborators{/collaborator}",
      "teams_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/teams",
      "hooks_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/hooks",
      "issue_events_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/issues/events{/number}",
      "events_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/events",
      "assignees_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/assignees{/user}",
      "branches_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/branches{/branch}",
      "tags_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/tags",
      "blobs_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/git/blobs{/sha}",
      "git_tags_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/git/tags{/sha}",
      "git_refs_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/git/refs{/sha}",
      "trees_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/git/trees{/sha}",
      "statuses_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/statuses/{sha}",
      "languages_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/languages",
      "stargazers_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/stargazers",
      "contributors_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/contributors",
      "subscribers_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/subscribers",
      "subscription_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/subscription",
      "commits_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/commits{/sha}",
      "git_commits_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/git/commits{/sha}",
      "comments_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/comments{/number}",
      "issue_comment_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/issues/comments{/number}",
      "contents_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/contents/{+path}",
      "compare_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/compare/{base}...{head}",
      "merges_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/merges",
      "archive_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/{archive_format}{/ref}",
      "downloads_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/downloads",
      "issues_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/issues{/number}",
      "pulls_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/pulls{/number}",
      "milestones_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/milestones{/number}",
      "notifications_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/notifications{?since,all,participating}",
      "labels_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/labels{/name}",
      "releases_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/releases{/id}",
      "deployments_url": "https://api.github.com/repos/research555/Cooperative-Protein-Dynamics/deployments",
      "created_at": "2020-12-27T13:27:35Z",
      "updated_at": "2020-12-30T14:58:26Z",
      "pushed_at": "2020-12-30T14:58:24Z",
      "git_url": "git://github.com/research555/Cooperative-Protein-Dynamics.git",
      "ssh_url": "git@github.com:research555/Cooperative-Protein-Dynamics.git",
      "clone_url": "https://github.com/research555/Cooperative-Protein-Dynamics.git",
      "svn_url": "https://github.com/research555/Cooperative-Protein-Dynamics",
      "homepage": null,
      "size": 6,
      "stargazers_count": 0,
      "watchers_count": 0,
      "language": "Python",
      "has_issues": true,
      "has_projects": true,
      "has_downloads": true,
      "has_wiki": true,
      "has_pages": false,
      "forks_count": 0,
      "mirror_url": null,
      "archived": false,
      "disabled": false,
      "open_issues_count": 0,
      "license": null,
      "forks": 0,
      "open_issues": 0,
      "watchers": 0,
      "default_branch": "main",
      "score": 1
    }
  ]
};
