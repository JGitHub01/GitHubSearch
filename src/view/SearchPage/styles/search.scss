$resultBackgroundColor: #F0F8FF;
.github-search {
  max-width: 900px;

  &__toolbar {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: flex-end;
    width: 100%;
  }

  &__sticky-header {
    display: flex;
    flex-direction: column;
    align-items: center;
    position: sticky;
    position: -webkit-sticky;
    top: 0;
    opacity: 1;
    width: 110%;
    background: white;
    padding-top: 16px;
    padding-bottom: 8px;
  }
  &__keyword {
    margin-bottom: 1rem;
    width: 90%;
  }

  &__count {
    display: flex;
    justify-content: space-around;
    width: 100%;
  }

  &-result {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 100%;

    &__list {
      display: flex;
      flex-direction: column;
      align-items: stretch;
      width: 100%;
      margin-top: 8px;
      margin-bottom: 8px;
      background-color: $resultBackgroundColor;

      .detail-link {
        text-decoration: none;
      }
      .repository {
        color: black;
        display: flex;
        flex-direction: row;
        justify-content: space-around;
        border-bottom: solid 1px #A9A9A9;
        margin-bottom: 8px;
        padding: 4px 8px;

        &:hover {
          background-color: darken($resultBackgroundColor, 10%);
        }
        &-description,
        &-name {
          display: flex;
          flex-direction: column;
          align-items: flex-start;

          &__label {
            color: #A9A9A9;
          }
        }

        &-name {
          width: 25%;
          margin-right: 8px;

          &__text {
            width: 100%;
            overflow: hidden;
            white-space: nowrap;
            text-overflow: ellipsis;
          }
        }

        &-description {
          width: 75%;
        }
      }
    }
  }

  &__sticky-footer {
    display: flex;
    justify-content: center;
    width: 110%;
    position: sticky;
    position: -webkit-sticky;
    bottom: 0;
    background-color: white;
    opacity: 1;
    padding-top: 8px;
    padding-bottom: 8px;

    .pagination {

      &-page--active {
        background-color: #6495ED;
      }
    }
  }

  &__language,
  &__sortby {
    display: flex;
    flex-direction: column;
    align-items: flex-start;

    select {
      padding: 2px 4px;
      margin-top: 2px;
    }
    @media(min-width: 639px) {
      flex-direction: row;
      align-items: center;

      label {
        margin-right: 8px;
      }

      select {
        padding: 4px 8px;
        margin-top: 0;
      }
    }
  }
}
