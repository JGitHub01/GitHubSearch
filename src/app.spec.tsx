import { render, waitFor, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import App from './app';
import { gitHubApiUrl } from 'lib/search';
import { search_result } from 'resources/tests-data/repository_data';

jest.mock('lib/responsive', () => {
  return {
    isScreenWidthAtLeast: () => true
  };
});

const server = setupServer(
  rest.get(gitHubApiUrl, (req, res, ctx) => {
    return res(ctx.json(search_result));
  })
);

describe('App', () => {
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());

  it('Should render the basic snapshot view', () => {
    const { getByTestId } = render(<App />);
    expect(getByTestId('search-page')).toMatchSnapshot();
  });
  it('Should initiate a search and display search result when a keyword is entered', async () => {
    jest.useFakeTimers();
    const { getByTestId, getAllByTestId } = render(<App />);
    const searchField = getByTestId('search-field');
    userEvent.type(searchField, 'tetris');
    expect(setTimeout).toBeCalledTimes(6);
    await act(() => { jest.runAllTimers(); return Promise.resolve(); });
    await waitFor(() => getAllByTestId('repository-item'));
    expect(getAllByTestId('repository-item')).toHaveLength(2);
    jest.useRealTimers();
  });
  it('Should initiate a search and display search result when a new language is selected', async () => {
    const { getAllByTestId, getByTestId, getByText } = render(<App />);
    const lanList = getByTestId('language-list');
    const phpOption = getByText('PhP');
    userEvent.selectOptions(lanList, phpOption);
    await waitFor(() => getAllByTestId('repository-item'));
    expect(getAllByTestId('repository-item')).toHaveLength(2);
  });
  it('Should initiate a search and display search result when sort by changes', async () => {
    const { getAllByTestId, getByTestId, getByText } = render(<App />);
    const sortbyList = getByTestId('sortby-list');
    const sortStars = getByText('Stars');
    userEvent.selectOptions(sortbyList, sortStars);
    await waitFor(() => getAllByTestId('repository-item'));
    expect(getAllByTestId('repository-item')).toHaveLength(2);
  });
  it('Should render error message for a failed search', async () => {
    server.use(
      rest.get(gitHubApiUrl, (req, res, ctx) => {
        return res(ctx.status(500))
      })
    );
    const { getByTestId, getByText } = render(<App />);
    const lanList = getByTestId('language-list');
    const phpOption = getByText('PhP');
    userEvent.selectOptions(lanList, phpOption);
    await waitFor(() => getByText('Trouble search GitHub', { exact: false }));
    expect(getByText('Trouble search GitHub', { exact: false })).toMatchSnapshot();
  });
  describe('Routing', () => {
    it('Should display the detail page when a repository item is clicked and go back the search page when "Back to search" is clicked',
      async () => {
        const { getAllByTestId, getByTestId, getByText } = render(<App />);
        const lanList = getByTestId('language-list');
        const phpOption = getByText('PhP');
        userEvent.selectOptions(lanList, phpOption);
        await waitFor(() => getAllByTestId('repository-item'));
        userEvent.click(getAllByTestId('repository-item')[0]);
        await waitFor(() => getByTestId('repository-detail'))
        expect(getByTestId('repository-detail')).toMatchSnapshot();
        userEvent.click(getByTestId('back-to-search'));
        await waitFor(() => getByTestId('search-page'));
        expect(getByTestId('search-page')).toMatchSnapshot();
      });
  });
});
