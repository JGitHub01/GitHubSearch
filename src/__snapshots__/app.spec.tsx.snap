// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Routing Should display the detail page when a repository item is clicked and go back the search page when "Back to search" is clicked 1`] = `
<div
  class="repository-detail"
  data-testid="repository-detail"
>
  <h2>
    Repository Detail
  </h2>
  <div
    class="repository-detail__body"
  >
    <button
      data-testid="back-to-search"
      type="button"
    >
      Back to search
    </button>
    <div
      class="repository"
    >
      <div
        class="repository-field"
      >
        <h3
          class="repository-field__label"
        >
          Name
        </h3>
        <div
          class="repository-field__text"
        >
          Tetrio-Rank-Predictor
        </div>
      </div>
      <div
        class="repository-field"
      >
        <h3
          class="repository-field__label"
        >
          Description
        </h3>
        <div
          class="repository-field__text"
        >
          Provides an approximate rank for Tetr.io's Tetra League based on APM, PPS, and VS-Score.
        </div>
      </div>
      <div
        class="repository-field"
      >
        <h3
          class="repository-field__label"
        >
          Language
        </h3>
        <div
          class="repository-field__text"
        >
          Python
        </div>
      </div>
      <div
        class="repository-field"
      >
        <h3
          class="repository-field__label"
        >
          Stars
        </h3>
        <div
          class="repository-field__text"
        >
          0
        </div>
      </div>
      <div
        class="repository-field"
      >
        <h3
          class="repository-field__label"
        >
          Owner
        </h3>
        <div
          class="repository-field__text"
        >
          chunkanglu
        </div>
      </div>
    </div>
    <a
      class="repository-detail__link"
      href="https://github.com/chunkanglu/Tetrio-Rank-Predictor"
      target="__blank"
    >
      Go to the repository
    </a>
  </div>
</div>
`;

exports[`App Routing Should display the detail page when a repository item is clicked and go back the search page when "Back to search" is clicked 2`] = `
<div
  class="github-search"
  data-testid="search-page"
>
  <h2>
    GitHub Repository Search
  </h2>
  <div
    class="github-search__toolbar"
  >
    <div
      class="github-search__sortby"
      data-testid="sortby-field"
    >
      <label
        for="sortby"
      >
        Sort by:
      </label>
      <select
        data-testid="sortby-list"
        name="sortby"
      >
        <option
          value=""
        >
          Best match
        </option>
        <option
          value="stars"
        >
          Stars
        </option>
      </select>
    </div>
    <div
      class="github-search__language"
      data-testid="search-language"
    >
      <label
        for="language"
      >
        Language:
      </label>
      <select
        data-testid="language-list"
        name="language"
      >
        <option
          value="c"
        >
          C
        </option>
        <option
          value="c#"
        >
          C#
        </option>
        <option
          value="java"
        >
          Java
        </option>
        <option
          value="javascript"
        >
          JavaScript
        </option>
        <option
          value="php"
        >
          PhP
        </option>
        <option
          value="python"
        >
          Python
        </option>
        <option
          value="ruby"
        >
          Ruby
        </option>
        <option
          value="shell"
        >
          Shell
        </option>
        <option
          value="typescript"
        >
          TypeScript
        </option>
      </select>
    </div>
  </div>
  <div
    class="github-search__sticky-header"
  >
    <input
      class="github-search__keyword"
      data-testid="search-field"
      type="text"
      value=""
    />
    <div
      class="github-search__count"
    >
      <span>
        Total: 40
      </span>
      <span>
        30/page
      </span>
    </div>
  </div>
  <div
    class="github-search-result"
    data-testid="search-result"
  >
    <div
      class="github-search-result__list"
    >
      <a
        class="detail-link"
        href="/detail"
      >
        <div
          class="repository"
          data-testid="repository-item"
        >
          <div
            class="repository-name"
          >
            <span
              class="repository-name__label"
            >
              Name:
            </span>
            <div
              class="repository-name__text"
            >
              Tetrio-Rank-Predictor
            </div>
          </div>
          <div
            class="repository-description"
          >
            <span
              class="repository-description__label"
            >
              Description:
            </span>
            <div
              class="repository-description__text"
            >
              Provides an approximate rank for Tetr.io's Tetra League based on APM, PPS, and VS-Score.
            </div>
          </div>
        </div>
      </a>
      <a
        class="detail-link"
        href="/detail"
      >
        <div
          class="repository"
          data-testid="repository-item"
        >
          <div
            class="repository-name"
          >
            <span
              class="repository-name__label"
            >
              Name:
            </span>
            <div
              class="repository-name__text"
            >
              Cooperative-Protein-Dynamics
            </div>
          </div>
          <div
            class="repository-description"
          >
            <span
              class="repository-description__label"
            >
              Description:
            </span>
            <div
              class="repository-description__text"
            >
              This is a simple program that maps the dynamics between LacI and TetR in a one of the first toggle switch systems created by Gardner et al (2000)
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
  <div
    class="github-search__sticky-footer"
  >
    <div
      class="pagination"
      data-testid="pagination"
    >
      <button
        class="pagination-left"
        data-testid="pagination-left"
        disabled=""
      >
        &lt;
      </button>
      <button
        class="pagination-page pagination-page--active"
      >
        1
      </button>
      <button
        class="pagination-page"
      >
        2
      </button>
      <button
        class="pagination-right"
        data-testid="pagination-right"
        disabled=""
      >
        &gt;
      </button>
    </div>
  </div>
</div>
`;

exports[`App Should render error message for a failed search 1`] = `
<div>
  Trouble search GitHub: 500
</div>
`;

exports[`App Should render the basic snapshot view 1`] = `
<div
  class="github-search"
  data-testid="search-page"
>
  <h2>
    GitHub Repository Search
  </h2>
  <div
    class="github-search__toolbar"
  >
    <div
      class="github-search__sortby"
      data-testid="sortby-field"
    >
      <label
        for="sortby"
      >
        Sort by:
      </label>
      <select
        data-testid="sortby-list"
        name="sortby"
      >
        <option
          value=""
        >
          Best match
        </option>
        <option
          value="stars"
        >
          Stars
        </option>
      </select>
    </div>
    <div
      class="github-search__language"
      data-testid="search-language"
    >
      <label
        for="language"
      >
        Language:
      </label>
      <select
        data-testid="language-list"
        name="language"
      >
        <option
          value="c"
        >
          C
        </option>
        <option
          value="c#"
        >
          C#
        </option>
        <option
          value="java"
        >
          Java
        </option>
        <option
          value="javascript"
        >
          JavaScript
        </option>
        <option
          value="php"
        >
          PhP
        </option>
        <option
          value="python"
        >
          Python
        </option>
        <option
          value="ruby"
        >
          Ruby
        </option>
        <option
          value="shell"
        >
          Shell
        </option>
        <option
          value="typescript"
        >
          TypeScript
        </option>
      </select>
    </div>
  </div>
  <div
    class="github-search__sticky-header"
  >
    <input
      class="github-search__keyword"
      data-testid="search-field"
      type="text"
      value=""
    />
    <div
      class="github-search__count"
    >
      <span>
        Total: 0
      </span>
      <span>
        30/page
      </span>
    </div>
  </div>
  <div
    class="github-search-result"
    data-testid="search-result"
  >
    <div>
      Please enter a keyword!
    </div>
  </div>
  <div
    class="github-search__sticky-footer"
  >
    <div
      class="pagination"
      data-testid="pagination"
    >
      <button
        class="pagination-left"
        data-testid="pagination-left"
        disabled=""
      >
        &lt;
      </button>
      <button
        class="pagination-right"
        data-testid="pagination-right"
        disabled=""
      >
        &gt;
      </button>
    </div>
  </div>
</div>
`;
